virtool/references/data.py

```diff
@@ -0,0 +1,658 @@
+                "manifest": manifest,
+                "ref_id": document["_id"],
+                "user_id": user_id,
+            }
+
+            task = await self.data.tasks.create(CloneReferenceTask, context=context)
+
+            document["task"] = {"id": task.id}
+
+        elif data.import_from:
+            if not await get_row(
+                self._pg, SQLUpload, ("name_on_disk", data.import_from)
+            ):
+                raise ResourceNotFoundError("File not found")
+
+            path = self._config.data_path / "files" / data.import_from
+
+            document = await virtool.references.db.create_import(
+                self._mongo,
+                self._pg,
+                settings,
+                data.name,
+                data.description,
+                data.import_from,
+                user_id,
+                data.data_type,
+                data.organism,
+            )
+
+            context = {
+                "created_at": document["created_at"],
+                "path": str(path),
+                "ref_id": document["_id"],
+                "user_id": user_id,
+            }
+
+            task = await self.data.tasks.create(ImportReferenceTask, context=context)
+
+            document["task"] = {"id": task.id}
+
+        elif data.remote_from:
+            try:
+                release = await virtool.github.get_release(
+                    self._config,
+                    self._client,
+                    data.remote_from,
+                    release_id=data.release_id,
+                )
+
+            except aiohttp.ClientConnectionError:
+                raise ResourceRemoteError("Could not reach GitHub")
+
+            except GitHubError as err:
+                if "404" in str(err):
+                    raise ResourceRemoteError(
+                        "Could not retrieve latest GitHub release"
+                    )
+
+                raise
+
+            release = format_release(release)
+
+            document = await virtool.references.db.create_remote(
+                self._mongo, settings, release, data.remote_from, user_id
+            )
+
+            context = {
+                "release": release,
+                "ref_id": document["_id"],
+                "created_at": document["created_at"],
+                "user_id": user_id,
+            }
+
+            task = await self.data.tasks.create(RemoteReferenceTask, context=context)
+
+            document["task"] = {"id": task.id}
+
+        else:
+            document = await virtool.references.db.create_document(
+                self._mongo,
+                settings,
+                data.name,
+                data.organism,
+                data.description,
+                data.data_type,
+                user_id=user_id,
+            )
+
+        document = await attach_computed(self._mongo, document)
+
+        await self._mongo.references.insert_one(document)
+
```

Comment:
This was done to include `handle` and `administrator` to the user and users fields before adding to the database as this may fix the bug where these values do not show in the find endpoint according to https://preview.virtool.ca/api/refs